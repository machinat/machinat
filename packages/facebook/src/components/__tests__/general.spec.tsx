import Sociably from '@sociably/core';
import Renderer from '@sociably/core/renderer';
import generalComponentDelegator from '../general';

const renderer = new Renderer('facebook', generalComponentDelegator);

test('render shallow element match snapshot', async () => {
  const results = await Promise.all(
    [
      <p>abc</p>,
      <b>important</b>,
      <i>italic</i>,
      <s>nooooo</s>,
      <u>underlined</u>,
      <code>foo.bar()</code>,
      <pre>foo.bar('hello world')</pre>,
      <br />,
    ].map((ele) => renderer.render(ele, null, null))
  );

  expect(results).toMatchSnapshot();
  expect(results.flat().map((seg) => seg?.value)).toMatchInlineSnapshot(`
    Array [
      "abc",
      "important",
      "italic",
      "nooooo",
      "underlined",
      "\`foo.bar()\`",
      "\`\`\`
    foo.bar('hello world')
    \`\`\`",
      "
    ",
    ]
  `);
});

test('render nested elements match snapshot', async () => {
  const segments = await renderer.render(
    <p>
      Mic test{' '}
      <code>
        Hello, <b>Luke Skywalker!</b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>I'm your</u> <s>FATHER</s> <code>droid</code>.
      </i>
      <br />
      <br />
      <pre>May the force be with you!</pre> Test over
    </p>,
    null,
    null
  );
  expect(segments).toMatchSnapshot();
  expect(segments?.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      "Mic test \`Hello, Luke Skywalker!\`
    You know what?
    I'm your FATHER \`droid\`.

    \`\`\`
    May the force be with you!
    \`\`\` Test over",
    ]
  `);
});

test('<p/> renders into individual text segment', async () => {
  const segments = await renderer.render(
    <>
      <p>foo</p>
      bar
      <p>baz</p>
    </>,
    null,
    null
  );

  expect(segments).toMatchSnapshot();
  expect(segments?.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      "foo",
      "bar",
      "baz",
    ]
  `);
});

test('throw if non-texual value received', async () => {
  const children = (
    <>
      foo
      <img />
      bar
    </>
  );

  await expect(
    renderer.render(<p>{children}</p>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <p/>"`
  );
  await expect(
    renderer.render(<b>{children}</b>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <b/>"`
  );
  await expect(
    renderer.render(<i>{children}</i>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <i/>"`
  );
  await expect(
    renderer.render(<s>{children}</s>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <s/>"`
  );
  await expect(
    renderer.render(<u>{children}</u>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <u/>"`
  );
  await expect(
    renderer.render(<code>{children}</code>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <code/>"`
  );
  await expect(
    renderer.render(<pre>{children}</pre>, null, null)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-textual node <img /> is placed in <pre/>"`
  );
});

test('render null if content is empty', async () => {
  const elements = [<b />, <i />, <s />, <u />, <p />, <code />, <pre />];
  for (const element of elements) {
    // eslint-disable-next-line no-await-in-loop
    await expect(renderer.render(element, null, null)).resolves.toEqual(null);
  }
});

test('media elements match snapshot', async () => {
  const segments = await renderer.render(
    <>
      <img src="http://avatar.my.bot" />
      <video src="http://vid.my.bot" />
      <audio src="http://sound.my.bot" />
      <file src="http://profile.my.bot" />
    </>,
    null,
    null
  );
  expect(segments).toMatchSnapshot();
  expect(segments?.map((seg) => seg.value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "url": "http://avatar.my.bot",
            },
            "type": "image",
          },
        },
      },
      Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "url": "http://vid.my.bot",
            },
            "type": "video",
          },
        },
      },
      Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "url": "http://sound.my.bot",
            },
            "type": "audio",
          },
        },
      },
      Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "url": "http://profile.my.bot",
            },
            "type": "file",
          },
        },
      },
    ]
  `);
});
