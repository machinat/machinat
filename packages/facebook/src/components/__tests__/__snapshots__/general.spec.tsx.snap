// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<p/> renders into individual text segment 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$::0",
    "type": "text",
    "value": "foo",
  },
  Object {
    "node": "bar",
    "path": "$::1",
    "type": "text",
    "value": "bar",
  },
  Object {
    "node": <p>
      baz
    </p>,
    "path": "$::2",
    "type": "text",
    "value": "baz",
  },
]
`;

exports[`media elements match snapshot 1`] = `
Array [
  Object {
    "node": <img
      src="http://avatar.my.bot"
    />,
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://avatar.my.bot",
          },
          "type": "image",
        },
      },
    },
  },
  Object {
    "node": <video
      src="http://vid.my.bot"
    />,
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://vid.my.bot",
          },
          "type": "video",
        },
      },
    },
  },
  Object {
    "node": <audio
      src="http://sound.my.bot"
    />,
    "path": "$::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://sound.my.bot",
          },
          "type": "audio",
        },
      },
    },
  },
  Object {
    "node": <file
      src="http://profile.my.bot"
    />,
    "path": "$::3",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://profile.my.bot",
          },
          "type": "file",
        },
      },
    },
  },
]
`;

exports[`render nested elements match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      Mic test
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        <u>
          I'm your
        </u>
         
        <s>
          FATHER
        </s>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       Test over
    </p>,
    "path": "$",
    "type": "text",
    "value": "Mic test \`Hello, Luke Skywalker!\`
You know what?
I'm your FATHER \`droid\`.

\`\`\`
May the force be with you!
\`\`\` Test over",
  },
]
`;

exports[`render shallow element match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <p>
        abc
      </p>,
      "path": "$",
      "type": "text",
      "value": "abc",
    },
  ],
  Array [
    Object {
      "node": <b>
        important
      </b>,
      "path": "$",
      "type": "text",
      "value": "important",
    },
  ],
  Array [
    Object {
      "node": <i>
        italic
      </i>,
      "path": "$",
      "type": "text",
      "value": "italic",
    },
  ],
  Array [
    Object {
      "node": <s>
        nooooo
      </s>,
      "path": "$",
      "type": "text",
      "value": "nooooo",
    },
  ],
  Array [
    Object {
      "node": <u>
        underlined
      </u>,
      "path": "$",
      "type": "text",
      "value": "underlined",
    },
  ],
  Array [
    Object {
      "node": <code>
        foo.bar()
      </code>,
      "path": "$",
      "type": "text",
      "value": "\`foo.bar()\`",
    },
  ],
  Array [
    Object {
      "node": <pre>
        foo.bar('hello world')
      </pre>,
      "path": "$",
      "type": "text",
      "value": "\`\`\`
foo.bar('hello world')
\`\`\`",
    },
  ],
  Array [
    Object {
      "node": <br />,
      "path": "$",
      "type": "text",
      "value": "
",
    },
  ],
]
`;
